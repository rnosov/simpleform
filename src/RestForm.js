/*
 * RestForm React Component
 *
 * Copyright Â© Roman Nosov 2016
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */
import React, { Component, PropTypes } from 'react';
import SimpleForm from './SimpleForm';
import scroller from 'react-scroll/lib/mixins/scroller';

const
  propTypes = {
    endpoint: PropTypes.string.isRequired,
    schema: PropTypes.object.isRequired,
    setStatus: PropTypes.func.isRequired,
    onResponseReceived: PropTypes.func.isRequired,
    onFormWillFetch: PropTypes.func,
    waitText: PropTypes.string,
    errorText: PropTypes.string,
    successText: PropTypes.string,
    welcomeText: PropTypes.string,
    scrollOrigin: PropTypes.string,
  },
  defaultProps = {
    scrollOrigin: 'content',
    waitText: 'Uploading form. Please wait ...',
    errorText: 'Houston, we have a problem!',
    successText: 'Your form has been submitted successfully',
    welcomeText: 'Welcome, please fill in the form below:',
  };

class RestForm extends Component {

  state = {
    msg: false,
    status: 'info',
    spinner: false,
  };
  isProcessing = false;

  jsonFetch(body) {
    return fetch(this.props.endpoint, {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'text/plain',
      },
      body: JSON.stringify(body),
    });
  }

  async handleApiCall(form) {
    if (this.props.onFormWillFetch)
      form = await this.props.onFormWillFetch(form);
    const response = await this.jsonFetch(form);
    if (response.status < 200 && response.status >= 300)
      throw new Error(response.statusText);
    return await this.props.onResponseReceived(response);
  }

  async handleSubmit(form) {
    if (this.isProcessing) return;
    this.isProcessing = true;
    this.props.setStatus('info', this.props.waitText, true);
    setTimeout( () => scroller.scrollTo(this.props.scrollOrigin, {
      duration: 500,
      delay: 0,
      smooth: true,
    }), 200);
    try {
      this.props.setStatus('success', await this.handleApiCall(form));
    }
    catch (err) {
      this.props.setStatus('danger', err.message || err);
    }
    finally {
      this.isProcessing = false;
    }
  }

  render() {
    if (typeof this.props.msg === 'string') {
      const goBack = <button onClick={ () => this.props.setStatus() } className="btn btn-outline-primary">Go back to the form</button>;
      return (
        <div className={`alert alert-` + this.props.status} role="alert">
          {this.props.status === 'danger' ?<h4 className="alert-heading">{this.props.errorText}</h4>:void 0}
          <p>{this.props.msg}</p>
          {goBack}
        </div>
      );
    }
    return(
      <div>
        <p>{this.props.welcomeText}</p>
        <SimpleForm
          schema={ this.props.schema }
          onSubmit={ ::this.handleSubmit }
          submitText={ this.props.submitText }
        />
      </div>
    );
  }

}

RestForm.propTypes = propTypes;
RestForm.defaultProps = defaultProps;

export default RestForm;
