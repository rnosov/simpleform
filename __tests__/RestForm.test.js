/*
 * RestForm Test Suite
 *
 * Copyright Â© Roman Nosov 2016
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import RestForm from '../src/RestForm';

describe('RestForm', () => {
  it('renders a initial view', () => {
    const form = mount(<RestForm endpoint="" schema={{}} setStatus={ () => void 0 } onResponseReceived={ () => void 0 } />);
    expect(form.html()).toMatchSnapshot();
  });
});
