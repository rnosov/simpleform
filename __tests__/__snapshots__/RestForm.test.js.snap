exports[`RestForm renders a initial view 1`] = `"<div><p>Welcome, please fill in the form below:</p><form class=\"m-x-1 text-xs-left\" style=\"overflow: hidden;\"><noscript>Please enable javascript in order to use this form.</noscript><div class=\"form-group row\"><div class=\"col-xs-12 col-sm-2\"></div><div class=\"col-xs-12 form-inline col-sm-8\"><div class=\"form-group has-danger\"><button type=\"submit\" class=\"btn btn-outline-primary\">Submit</button><div class=\"text-muted form-control-static\">&nbsp;&nbsp; </div><!-- react-empty: 11 --></div></div></div></form></div>"`;
